### PR TITLE
[RPC] add sendmany

### PR DESCRIPTION
**Pull request by mikedennis**
Part of #2758 

Note: subtractfeefrom is currently not supported because it is not yet supported in WalletTransactionHandler

https://bitcoincore.org/en/doc/0.16.1/rpc/wallet/sendmany/

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs [hunk 1]
```diff
@@ -261,6 +261,73 @@ public UnspentCoinModel[] ListUnspent(int minConfirmations = 1, int maxConfirmat
...
+        [ActionDescription("Creates and broadcasts a transaction which sends outputs to multiple addresses.")]
+        public async Task<uint256> SendManyAsync(string fromAccount, string addressesJson, int minConf = 1, string comment = null, string subtractFeeFromJson = null, bool isReplaceable = false, int? confTarget = null, string estimateMode = "UNSET")
+        {
+            var accountReference = this.GetAccount();
+           
+            // Optional list of addresses to subtract fees from.
+            IEnumerable<BitcoinAddress> subtractFeeFromAddresses = null;            
+            if (!string.IsNullOrEmpty(subtractFeeFromJson))
+            {
+                subtractFeeFromAddresses = JsonConvert.DeserializeObject<List<string>>(subtractFeeFromJson).Select(i => BitcoinAddress.Create(i, this.fullNode.Network));
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Perhaps wrap this in a try catch if the Json is malformed and return null?

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs [hunk 2]
```diff
@@ -261,6 +261,73 @@ public UnspentCoinModel[] ListUnspent(int minConfirmations = 1, int maxConfirmat
             return unspentCoins.ToArray();
         }
 
+        [ActionName("sendmany")]
+        [ActionDescription("Creates and broadcasts a transaction which sends outputs to multiple addresses.")]
+        public async Task<uint256> SendManyAsync(string fromAccount, string addressesJson, int minConf = 1, string comment = null, string subtractFeeFromJson = null, bool isReplaceable = false, int? confTarget = null, string estimateMode = "UNSET")
+        {
+            var accountReference = this.GetAccount();
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: You can declare this variable closer to where it is used.

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Also, can someone submit an empty list?

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs [hunk 3]
```diff
@@ -261,6 +261,73 @@ public UnspentCoinModel[] ListUnspent(int minConfirmations = 1, int maxConfirmat
...
+            if (!string.IsNullOrEmpty(subtractFeeFromJson))
+            {
+                subtractFeeFromAddresses = JsonConvert.DeserializeObject<List<string>>(subtractFeeFromJson).Select(i => BitcoinAddress.Create(i, this.fullNode.Network));
+            }
+
+            // Outputs addresses are keyvalue pairs of address, amount. Translate to Receipient list.
+            IEnumerable<Recipient> recipients = null;
+            if (!string.IsNullOrEmpty(addressesJson))
+            {
+                recipients = JsonConvert.DeserializeObject<Dictionary<string, decimal>>(addressesJson).Select(address => new Recipient
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Same here perhaps protect the deserialization with a try catch.

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs [hunk 4]
```diff
@@ -261,6 +261,73 @@ public UnspentCoinModel[] ListUnspent(int minConfirmations = 1, int maxConfirmat
...
+            if (estimateMode.Equals("ECONOMICAL", StringComparison.InvariantCultureIgnoreCase))
+                context.FeeType = FeeType.Low;
+            else if (estimateMode.Equals("CONSERVATIVE", StringComparison.InvariantCultureIgnoreCase))
+                context.FeeType = FeeType.High;
+
+            try
+            { 
+                // Log warnings for currently unsupported parameters.
+                if (!string.IsNullOrEmpty(comment))
+                    this.logger.LogWarning("'comment' parameter is currently unsupported. Ignored.");
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Change to:

this.logger.LogWarning("'{0}' parameter is currently unsupported. Ignored.", nameof(comment));

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs [hunk 5]
```diff
@@ -261,6 +261,73 @@ public UnspentCoinModel[] ListUnspent(int minConfirmations = 1, int maxConfirmat
...
+                if (isReplaceable)
+                    this.logger.LogWarning("'replaceable' parameter is currently unsupported. Ignored.");
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Use `nameof` here as well.

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs [hunk 6]
```diff
@@ -261,6 +261,73 @@ public UnspentCoinModel[] ListUnspent(int minConfirmations = 1, int maxConfirmat
...
+                if (confTarget != null)
+                    this.logger.LogWarning("'conf_target' parameter is currently unsupported. Ignored.");
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Use `nameof` here as well.

File: src/Stratis.Bitcoin.IntegrationTests/RPC/RPCTestsMutable.cs [hunk 7]
```diff
@@ -181,5 +183,65 @@ public void TestRpcSendToAddressCantSpendWhenLocked()
                 Assert.Throws<RPCException>(() => rpcClient.SendToAddress(aliceAddress, Money.Coins(1.0m)));
             }
         }
+
+        [Fact]
+        public void TestRpcSendManyWithValidDataIsSuccessful()
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.IntegrationTests/RPC/RPCTestsMutable.cs]: Perhaps add some tests for malformed json or where the send fail?

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs (hunk 5 above)
Comment (mikedennis) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: replaceable is the name of the rpc parameter in rpc docs. The parameter name in this method doesn't matter as the order the parameters are passed is how it determines which one is which.

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs (hunk 6 above)
Comment (mikedennis) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Ditto here, rpc parameter name is not the same as method parameter name

**Comment by mikedennis** [comment]
@fassadlr No can 'o worms thankfully! Added some more error handling and a test as per your suggestions.

File: src/Stratis.Bitcoin.IntegrationTests/RPC/RPCTestsMutable.cs [hunk 8]
```diff
@@ -181,5 +184,91 @@ public void TestRpcSendToAddressCantSpendWhenLocked()
...
+                txid.Should().Be(tx.GetHash());                
+
+                // Check the output is the right amount.
+                var aliceCoins = tx.Outputs.AsCoins().First(c => c.ScriptPubKey == aliceAddress.ScriptPubKey);
+                aliceCoins.Amount.Should().Be(Money.Coins(coinsForAlice));
+                
+                var bobCoins = tx.Outputs.AsCoins().First(c => c.ScriptPubKey == bobAddress.ScriptPubKey);
+                bobCoins.Amount.Should().Be(Money.Coins(coinsForBob));
+
+                // TODO: Enable test below once WalletTransactionHandler supports recepients with SubtractFeeFromAmount=true
```
Comment (dangershony) [inline src/Stratis.Bitcoin.IntegrationTests/RPC/RPCTestsMutable.cs]: @mikedennis can you create an issue to add support to this (unless there is one already)

**Comment by fassadlr** [comment]
Hi Mike,

I've implemented some extra checks as per the core source code:

https://github.com/bitcoin/bitcoin/blob/86e0a33f5c382513d5179e3fdf158baf952d7e2f/src/wallet/rpcwallet.cpp#L785

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs [hunk 9]
```diff
@@ -265,72 +265,100 @@ public UnspentCoinModel[] ListUnspent(int minConfirmations = 1, int maxConfirmat
...
+                    ScriptPubKey = recipientAddress,
+                    Amount = amount,
+                    SubtractFeeFromAmount = subtractFeeFromAddresses == null ? false : subtractFeeFromAddresses.Contains(BitcoinAddress.Create(address.Key, this.fullNode.Network))
+                };
+
+                recipients.Add(recipient);
             }
 
             var accountReference = this.GetAccount();
+            var spendable = this.walletManager.GetSpendableTransactionsInAccount(accountReference, minConf);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Is this check not made inside the create trx?

Comment (mikedennis) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Double checked, negative amount will throw an ArgumentOutOfRange exception in wallet. Duplicate recipient is an interesting one, I don't think the wallet cares about that as it's valid to send to the same address more than once (a bit dumb but valid). So I guess we can do this check in the RPC if core does.

Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: > Double checked, negative amount will throw an ArgumentOutOfRange exception in wallet. 

OK sweet, I will remove this check then 👍

**Comment by fassadlr** [comment]
@mikedennis let me if you are happy and we can push this once build passes 👍

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs [hunk 10]
```diff
@@ -253,21 +253,126 @@ public UnspentCoinModel[] ListUnspent(int minConfirmations = 1, int maxConfirmat
...
+            var recipients = new List<Recipient>();
+            foreach (var address in addresses)
+            {
+                // Check for duplicate recipients
+                var recipientAddress = BitcoinAddress.Create(address.Key, this.fullNode.Network).ScriptPubKey;
+                if (recipients.Any(r => r.ScriptPubKey == recipientAddress))
+                    throw new RPCServerException(RPCErrorCode.RPC_INVALID_PARAMETER, string.Format("Invalid parameter, duplicated address: {0}.", recipientAddress));
+
+                var amount = Money.Coins(address.Value);
+                if (amount <= 0)
```
Comment (mikedennis) [inline src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: This is checked when the transaction is built and ArgumentOutOfRange exception is thrown by wallet. Could potentially replace this check with a catch block for this exception.

File: src/Stratis.Bitcoin.IntegrationTests/RPC/RPCTestsMutable.cs (hunk 8 above)
Comment (mikedennis) [reply src/Stratis.Bitcoin.IntegrationTests/RPC/RPCTestsMutable.cs]: Done. Added #3023 issue

**Comment by mikedennis** [comment]
@fassadlr Looks good, just one suggestion that you may be potentially able to replace the one check with just an exception handler.

File: src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs (hunk 10 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs]: Ok yeah I have now also removed this check 👍

**Pull request merged**